### PR TITLE
feat: add buffering layer to BulkWriter

### PR DESCRIPTION
**Pull request by thebrianchen**
Porting from [node](https://github.com/googleapis/nodejs-firestore/pull/1470)

### CODE REVIEW

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 1]
```diff
@@ -660,10 +691,72 @@ public Void apply(BulkWriterOperation operation) {
...
+                  pendingOpsCount++;
+                  sendOperationLocked(enqueueOperationOnBatchCallback, operation);
+                }
+              }
+            });
+      }
     }
 
-    return operation.getFuture();
+    ApiFuture<WriteResult> transformedFuture =
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:727]: A slightly less generic name would do wonders here :)

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java [hunk 2]
```diff
@@ -660,10 +691,72 @@ public Void apply(BulkWriterOperation operation) {
...
+        ApiFutures.transformAsync(
+            operation.getFuture(),
+            new ApiAsyncFunction<WriteResult, WriteResult>() {
+              public ApiFuture<WriteResult> apply(WriteResult result) throws Exception {
+                pendingOpsCount--;
+                processBufferedOperations();
+                return ApiFutures.immediateFuture(result);
+              }
+            });
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:736]: Does this not need an executor for the non-deprecated version?

**Review by schmidt-sebastian — APPROVED** [review]
yay!

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 2 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:736]: added executor, thanks for catching!

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java (hunk 1 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java:727]: renamed to `processedOperationFuture`.

**Pull request merged**